backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,336 @@
+  private final Map<String, BackendDetail> backendDetailLookup = new ConcurrentHashMap<>();
+  private final Map<Recorder.ProcessGroup, String> processGroupToBackendLookup = new ConcurrentHashMap<>();
+  private final PriorityQueue<BackendDetail> availableBackendsByPriority;
+  private final ReentrantLock backendAssignmentLock = new ReentrantLock();
+
+  private ZookeeperBasedBackendAssociationStore(Vertx vertx, CuratorFramework curatorClient, String backendAssociationPath,
+                                                int reportingFrequencyInSeconds, int maxAllowedSkips,
+                                                Comparator<BackendDetail> backendPriorityComparator)
+      throws Exception {
+    this.vertx = vertx;
+    this.curatorClient = curatorClient;
+    this.backendAssociationPath = backendAssociationPath;
+    this.reportingFrequencyInSeconds = reportingFrequencyInSeconds;
+    this.maxAllowedSkips = maxAllowedSkips;
+
+    List<BackendDetail> existingBackendsInZookeeper = loadDataFromZookeeper();
+    this.availableBackendsByPriority = new PriorityQueue<>(backendPriorityComparator);
+    this.availableBackendsByPriority.addAll(existingBackendsInZookeeper);
+
+    for(BackendDetail backendDetail: existingBackendsInZookeeper) {
+      this.backendDetailLookup.put(backendDetail.getBackendIPAddress(), backendDetail);
+      for(Recorder.ProcessGroup processGroup: backendDetail.getAssociatedProcessGroups()) {
+        if (this.processGroupToBackendLookup.putIfAbsent(processGroup, backendDetail.getBackendIPAddress()) != null) {
+          throw new IllegalStateException(String.format("Backend mapping already exists for process group=%s", processGroup));
+        }
+      }
+    }
+  }
+
+  @Override
+  public Future<Recorder.ProcessGroups> reportBackendLoad(BackendDTO.LoadReportRequest payload) {
+    String backendIPAddress = payload.getIp();
+    Future<Recorder.ProcessGroups> result = Future.future();
+    vertx.executeBlocking(future -> {
+      try {
+        BackendDetail backendDetail = backendDetailLookup.computeIfAbsent(backendIPAddress, (key) -> {
+          try {
+            BackendDetail updatedBackendDetail = new BackendDetail(key, reportingFrequencyInSeconds, maxAllowedSkips);
+            String zNodePath = getZNodePathForBackend(key);
+            ZookeeperUtil.writeZNode(curatorClient, zNodePath, BackendDetail.serializeProcessGroups(updatedBackendDetail.getAssociatedProcessGroups()), true);
+            return updatedBackendDetail;
+          } catch (Exception ex) {
+            throw new RuntimeException(ex);
+          }
+        });
+
+        /**
+         * If backend was defunct earlier, acquire lock on available backend queue and add this backend there along with reporting load
+         * If backend was available, don't incur cost of acquiring lock since backend will already be in queue, which is the case mostly. Report load and get out
+         */
+        boolean wasDefunct = backendDetail.isDefunct();
```

Comment:
isDefunct method accepts system nanotime and uses updateGet semantics of atomic long to update atomically